cartography/intel/gcp/compute.py

```diff
@@ -0,0 +1,170 @@
+# Google Compute Engine API-centric functions
+# https://cloud.google.com/compute/docs/concepts
+from googleapiclient.discovery import HttpError
+import json
+import logging
+
+from cartography.util import run_cleanup_job
+
+logger = logging.getLogger(__name__)
+
+
+def _get_error_reason(http_error):
+    """
+    Helper function to get an error reason out of the googeapiclient's HttpError object
+    This function copies the structure of
+    https://github.com/googleapis/google-api-python-client/blob/master/googleapiclient/errors.py#L46.
+    At the moment this is the best way we know of to extract the HTTP failure reason.
+    Additionally, see https://github.com/googleapis/google-api-python-client/issues/662.
+    :param http_error: The googleapi HttpError object
+    :return: The error reason as a string
+    """
+    try:
+        data = json.loads(http_error.content.decode('utf-8'))
+        if isinstance(data, dict):
+            first_error = data
+        elif isinstance(data, list) and len(data) > 0:
+            first_error = data[0]
+        reason = first_error['error']['errors'][0]['reason']
+    except (ValueError, KeyError, TypeError):
+        logger.error("Could not parse HttpError object, returning `cartography.ExceptionNotParsable` as reason."
```

Comment:
As the code is currently written this error will be logged three (four if you count stdout) times with different messages each time. First here, second when the return value of this function is checked by `get_zones_in_project` (line 79 in this file as of this comment), third when the exception raised in `get_zones_in_project` is caught by `Sync.run` and then fourth when the program crashes due to the uncaught exception.

We need to think harder about how to handle this situation. Log-spam is the immediate issue but there's a larger question about exceptions and how they effect the execution of the sync in general which we have discussed before.

Leaving this comment to spark a broader discussion, but the immediate ask is to address the log-spam somehow. We can figure out a good way of approaching that together.

Reply:
Got it, I like the general principle where we re-raise all exceptions unless we know how to actually handle them but I missed that here.  My latest push does this I think